Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -0,0 +1,275 @@
+"""
+Python version 3.7.0
+2.4 - Partition
+Write code to partition a linked list around a value x,
+such that all nodes less than x come before all nodes
+greater than or equal to x. If x is contained within
+the list, the values of x only need to be after
+the elements less than x. The partition element x
+can appear anywhere in the 'right partition'; it does
+not need to appear between the left and right
+partitions.
+EXAMPLE
+Input:  3 -> 5 -> 8 -> 5 -> 10 -> 2 -> 1 [pivot = 5]
+Result: 3 -> 1 -> 2 -> 10 -> 5 -> 5 -> 8
+"""
+import unittest
+from typing import Optional
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        self.size = 0
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, e) -> None:
+        if isinstance(e, int):
+            self._append_num(e)
+        elif isinstance(e, Node) or e is None:
+            self._append_node(e)
+
+    def _append_num(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def _append_node(self, n: Optional[Node] = None) -> None:
```

Comment:
argh, should have caught this earlier: you now don't need this to be `Optional`!

Reply:
going to update rn

Reply:
updated. btw, thank for taking the time to review my code.  I feel like I'm becoming a better programmer than before.  I really appreciate it!